src/lib.rs

```diff
@@ -114,36 +121,25 @@ pub(crate) const PRESERVED_VALUE_MAGIC: &str = "1fc430ca-5b7f-4295-92de-33cf2b14
 #[serde(rename = "1fc430ca-5b7f-4295-92de-33cf2b145d38")]
 struct PreservedValueWrapper(u32);
 
-impl<'de, T: JsCast> serde::Deserialize<'de> for PreservedValue<T> {
-    fn deserialize<D>(deserializer: D) -> std::result::Result<Self, D::Error>
-    where
-        D: serde::Deserializer<'de>,
-    {
-        use serde::de::Error;
-
-        PreservedValueWrapper::deserialize(deserializer).and_then(|wrapper| {
-            // When used with our deserializer this unsafe is correct, because the
-            // deserializer just converted a JsValue into_abi.
-            // With other deserializers, this may be incorrect but it shouldn't be UB
-            // because JsValues are represented using indices into a JS-side (i.e.
-            // bounds-checked) array.
-            let val: JsValue = unsafe { FromWasmAbi::from_abi(wrapper.0) };
-            val.dyn_into()
-                .map(|val| PreservedValue(val))
-                .map_err(|e| D::Error::custom(format!("incompatible JS value {e:?}")))
-        })
+impl<T: JsCast + Clone> From<PreservedValue<T>> for PreservedValueWrapper {
+    fn from(val: PreservedValue<T>) -> Self {
+        PreservedValueWrapper(val.0.into().into_abi())
     }
 }
 
-impl<T: JsCast> serde::Serialize for PreservedValue<T> {
-    fn serialize<S>(&self, serializer: S) -> std::result::Result<S::Ok, S::Error>
-    where
-        S: serde::Serializer,
-    {
-        // Since we don't own `self` we need to clone it. Otherwise serializing a JsValue
-        // will produce a second JsValue pointing to the same index in the wasm-bindgen heap,
-        // and whichever one is dropped first will leave the other one dangling.
-        let idx = self.0.as_ref().clone().into_abi();
-        PreservedValueWrapper(idx).serialize(serializer)
+impl<T: JsCast + Clone> TryFrom<PreservedValueWrapper> for PreservedValue<T> {
```

Comment:
Where is `: Clone` bound is used / is it still?

Reply:
It's required for `serde(into)`. (Since `Serialize` takes a borrow, maybe there should be something like `serde(asref)`, but I couldn't find anything like that.)

Reply:
Hmm that's unfortunate. Maybe in that case it is better to implement `Serialize` by hand after all. Sorry for misleading.

Reply:
Or rather, maybe then use `serialize_with`/`deserialize_with` on the field itself. That way you still get the wrapper derived, and only the field needs special handling.

Reply:
Ah yes, `with` seems to be better than into/from. This gets rid of the `Clone` bound and makes the error-handling better.

(Also, no worries about the back-and-forth. It's been educational!)